src/metabase/api/dashboard.clj

```diff
@@ -539,24 +513,67 @@
                                          (assoc mapping :card-id (get dashcard-id->card-id dashcard-id)))]
     (check-parameter-mapping-permissions new-mappings)))
 
+(mu/defn ^:private classify-changes :- [:map
+                                        [:to-create [:sequential [:map [:id neg-int?]]]]
+                                        [:to-update [:sequential [:map [:id ms/PositiveInt]]]]
+                                        [:to-delete [:sequential [:map [:id ms/PositiveInt]]]]]
+  [current-changes :- [:sequential [:map [:id ms/PositiveInt]]]
+   new-changes     :- [:sequential [:map [:id int?]]]]
+  (let [current-dashcard-ids          (set (map :id current-changes))
+        update-or-delete-dashcard-ids (->> new-changes
+                                          (map :id)
+                                          (filter pos?)
+                                          set)
+        to-update (filter #(and (current-dashcard-ids (:id %))
+                                (update-or-delete-dashcard-ids (:id %)))
+                          new-changes)
+        to-delete (filter #(not (update-or-delete-dashcard-ids (:id %)))
+                           current-changes)
+        to-create (filter #(neg? (:id %)) new-changes)]
+    {:to-update to-update
+     :to-delete to-delete
+     :to-create to-create}))
+
+(defn- create-cards!
+  [dashboard cards]
+  (for [{:keys [card_id]} cards
+        :when  card_id]
+    (api/check-not-archived (api/read-check Card card_id)))
+  (check-parameter-mapping-permissions (flatten (for [{:keys [card_id parameter_mappings]} cards]
+                                                  (for [mapping parameter_mappings]
+                                                    (assoc mapping :card-id card_id)))))
```

Comment:
```suggestion
  (check-parameter-mapping-permissions (for [{:keys [card_id parameter_mappings]} cards
                                             mapping parameter_mappings]
                                         (assoc mapping :card-id card_id)))
```

Reply:
awesome, forgot that we could have nested for destructring

Reply:
0d7da010b2